bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieWatcherImpl.java

```diff
@@ -173,6 +173,22 @@ public BookieAddressResolver getBookieAddressResolver() {
         }
     }
 
+    /**
+     * Determine if a bookie should be considered unavailable.
+     * This does not require a network call because this class
+     * maintains a current view of readonly and writable bookies.
+     * An unavailable bookie is one that is neither read only nor
+     * writable.
+     *
+     * @param id
+     *          Bookie to check
+     * @return whether or not the given bookie is unavailable
+     */
+    @Override
+    public boolean isBookieUnavailable(BookieId id) {
+        return !readOnlyBookies.contains(id) && !writableBookies.contains(id);
```

Comment:
There is any need to add more concurrency controls here. `readOnlyBookies` and `writableBookies` are already declared as `volatile`, which will ensure that this method call gets the latest values for these variables.